tests/unit/test_adjoints.py

```diff
@@ -0,0 +1,218 @@
+import pytest
+import pickle
+from firedrake import *
+from firedrake.adjoint import *
+from checkpoint_schedules import SingleMemoryStorageSchedule, SingleDiskStorageSchedule
+import numpy as np
+from pathlib import Path
+
+
+def tape_generation_staggered_solves(scheduler):
+
+    continue_annotation()
+
+    # get tape
+    tape = get_working_tape()
+    tape.clear_tape()
+    ret = {}
+
+    if scheduler is not None:
+        if isinstance(scheduler, SingleDiskStorageSchedule):
+            enable_disk_checkpointing()
+        tape.enable_checkpointing(scheduler)
+
+    mesh = checkpointable_mesh(UnitSquareMesh(1, 1)) if isinstance(scheduler, SingleDiskStorageSchedule) else UnitSquareMesh(1, 1)
+
+    V = FunctionSpace(mesh, "CG", 1)
+
+    u_0 = Function(V).assign(1.0)
+    u = Function(V).assign(u_0)
+    r = Function(V)
+
+    for i in tape.timestepper(iter(range(10))):
+        if i % 3 == 0:
+            r.project(1.01 * u)
+        u.project(r * u)
+
+    J = assemble((u) ** 2 * dx)
+
+    pause_annotation()
+
+    reduced_functional = ReducedFunctional(J, Control(u_0))
+
+    # Storing the diagnostics
+    ret["using scheduler"] = scheduler.__class__.__name__ if scheduler is not None else "None"
+    ret["J1"] = reduced_functional(u_0)
+    ret["dJdm1"] = reduced_functional.derivative().dat.data_ro.copy()
+    return ret
+
+
+def tape_generation_control_invariant_assign(scheduler):
+    """ See bellow test_control_invariant_assign for references """
+    continue_annotation()
+    tape = get_working_tape()
+    tape.clear_tape()
+
+    if isinstance(scheduler, SingleDiskStorageSchedule):
+        enable_disk_checkpointing()
+
+    if scheduler is not None:
+        tape.enable_checkpointing(scheduler)
+
+    mesh = UnitSquareMesh(1, 1)
+
+    if isinstance(scheduler, SingleDiskStorageSchedule):
+        mesh = checkpointable_mesh(mesh)
```

Comment:
Unify structure with above, or maybe even pull the whole prologue of the tape generation functions out and return a `mesh` compatible with the scheduler after enabling annotation and checkpointing.